// @patternfly-labs/react-form-wizard/lib/wizards/Routes.js
export var RouteE;
(function (RouteE) {
    RouteE["Home"] = ".";
    RouteE["ArgoCD"] = "?route=argo-cd";
    RouteE["Ansible"] = "?route=ansible";
    RouteE["Application"] = "?route=application";
    RouteE["Cluster"] = "?route=cluster";
    RouteE["Credentials"] = "?route=credentials";
    RouteE["Placement"] = "?route=placement";
    RouteE["CreatePlacement"] = "?route=create-placement";
    RouteE["CreatePlacementRule"] = "?route=create-placement-rule";
    RouteE["EditPlacement"] = "?route=edit-placement";
    RouteE["EditPlacementRule"] = "?route=edit-placement-rule";
    RouteE["EditPlacements"] = "?route=edit-placements";
    RouteE["Policy"] = "?route=policy";
    RouteE["PolicySet"] = "?route=policy-set";
    RouteE["CreatePolicy"] = "?route=create-policy";
    RouteE["EditPolicyLimitClusterAdmin"] = "?route=edit-policy-limit-cluster-admin";
    RouteE["EditPolicyRole"] = "?route=edit-policy-role";
    RouteE["EditPolicyRoleBinding"] = "?route=edit-policy-role-binding";
    RouteE["EditPolicyComplianceOperatorInstall"] = "?route=edit-policy-compliance-operator-install";
    RouteE["EditPolicyComplianceOperatorCisScan"] = "?route=edit-policy-compliance-operator-cis-scan";
    RouteE["EditPolicyComplianceOperatorE8Scan"] = "?route=edit-policy-compliance-operator-e8-scan";
    RouteE["EditPolicyGatekeeperOperatorDownstream"] = "?route=edit-policy-compliance-operator-downstream";
    RouteE["EditPolicyNamespace"] = "?route=edit-policy-namespace";
    RouteE["EditPolicyPod"] = "?route=edit-policy-pod";
    RouteE["EditPolicyCertificate"] = "?route=edit-policy-certificate";
    RouteE["EditPolicyEtcdEncryption"] = "?route=edit-policy-etcd-encryption";
    RouteE["EditPolicyLimitMemory"] = "?route=edit-policy-limit-memory";
    RouteE["EditPolicyImageManifestVuln"] = "?route=edit-policy-image-manifect-vuln";
    RouteE["EditPolicyPsp"] = "?route=edit-policy-psp";
    RouteE["EditPolicyScc"] = "?route=edit-policy-scc";
    RouteE["CreatePolicySet"] = "?route=create-policy-set";
    RouteE["EditPolicySet1"] = "?route=edit-policy-set-1";
    RouteE["EditPolicySet2"] = "?route=edit-policy-set-2";
    RouteE["EditPolicySet3"] = "?route=edit-policy-set-3";
    RouteE["EditPolicySet4"] = "?route=edit-policy-set-4";
    RouteE["EditPolicySet5"] = "?route=edit-policy-set-5";
    RouteE["EditPolicySet6"] = "?route=edit-policy-set-6";
    RouteE["EditPolicySet7"] = "?route=edit-policy-set-7";
    RouteE["EditPolicySet8"] = "?route=edit-policy-set-8";
    RouteE["ROSA"] = "?route=rosa";
    RouteE["App"] = "?route=app";
    RouteE["Inputs"] = "?route=inputs";
    RouteE["Wizards"] = "?route=wizards";
})(RouteE || (RouteE = {}));
