// @patternfly-labs/react-form-wizard/lib/src/components/YamlEditor.js
import { jsx as _jsx, jsxs as _jsxs } from "react/jsx-runtime";
import { Button } from '@patternfly/react-core';
import { CheckIcon, CopyIcon } from '@patternfly/react-icons';
import { useEffect, useState } from 'react';
import YAML from 'yaml';
const color = {
    background: 'rgb(21, 21, 21)',
    divider: 'rgb(212, 212, 212)',
    colon: 'rgb(212, 212, 212)',
    variable: 'rgb(115, 188, 247)',
    value: 'rgb(240, 171, 0)',
};
export function YamlToObject(yaml, isYamlArray) {
    if (isYamlArray === true) {
        return YAML.parseAllDocuments(yaml, { prettyErrors: true }).map((doc) => doc.toJSON());
    }
    else if (isYamlArray === false) {
        return YAML.parse(yaml, { prettyErrors: true });
    }
    try {
        return YAML.parse(yaml, { prettyErrors: true });
    }
    catch {
        return YAML.parseAllDocuments(yaml, { prettyErrors: true }).map((doc) => doc.toJSON());
    }
}
export function ObjectToYaml(data, isYamlArray) {
    if (isYamlArray) {
        return data.map((doc) => YAML.stringify(doc)).join('---\n');
    }
    else {
        return YAML.stringify(data);
    }
}
export function YamlEditor(props) {
    const [hasFocus, setHasFocus] = useState(false);
    const [, setError] = useState('');
    const [errorLine, setErrorLine] = useState(-1);
    const [yaml, setYaml] = useState(() => {
        return ObjectToYaml(props.data, props.isYamlArray) ?? '';
    });
    useEffect(() => {
        if (!hasFocus) {
            setYaml(ObjectToYaml(props.data, props.isYamlArray));
            setError('');
            setErrorLine(-1);
        }
    }, [props.data, hasFocus, props.isYamlArray]);
    const [copied, setCopied] = useState(false);
    return (_jsxs("div", { style: { display: 'flex', flexDirection: 'column', flexGrow: 1, maxHeight: '100%' }, children: [_jsx("div", { style: {
                    display: 'flex',
                    borderBottom: 'thin solid #ffffff30',
                    justifyContent: 'end',
                    backgroundColor: '#ffffff18',
                    color: 'white',
                    alignItems: 'center',
                }, children: _jsx(Button, { variant: "plain", onClick: () => {
                        void navigator.clipboard.writeText(yaml);
                        setCopied(true);
                        setTimeout(() => setCopied(false), 2000);
                    }, tabIndex: -1, children: copied ? _jsx(CheckIcon, { color: "var(--pf-global--success-color--100)" }, void 0) : _jsx(CopyIcon, { color: "white" }, void 0) }, void 0) }, void 0), _jsx("div", { style: { display: 'flex', flexGrow: 1, overflow: 'auto' }, children: _jsx("pre", { style: { display: 'flex', flexGrow: 1, justifySelf: 'stretch', position: 'relative', padding: 24 }, onFocus: () => setHasFocus(true), onBlur: () => setHasFocus(false), children: _jsxs("small", { children: [yaml.split('\n').map((line, index) => {
                                const backgroundColor = index === errorLine ? '#F203' : undefined;
                                if (line === '---') {
                                    return (_jsx("div", { style: { color: color.divider, backgroundColor }, children: line }, index));
                                }
                                const parts = line.split(':');
                                if (parts[0] === '')
                                    return _jsx("div", { children: "\u00A0" }, index);
                                if (parts.length === 1) {
                                    return (_jsx("div", { style: { color: color.variable, backgroundColor }, children: parts[0] }, index));
                                }
                                return (_jsxs("div", { style: { color: color.variable, backgroundColor }, children: [parts[0], _jsx("span", { style: { color: color.colon }, children: ":" }, void 0), _jsx("span", { style: { color: color.value }, children: parts.slice(1).join(':') }, void 0)] }, index));
                            }), _jsx("textarea", { id: "yaml-editor", style: {
                                    position: 'absolute',
                                    top: 0,
                                    left: 0,
                                    height: '100%',
                                    width: '100%',
                                    padding: 24,
                                    border: 0,
                                    backgroundColor: 'transparent',
                                    whiteSpace: 'pre',
                                    overflowWrap: 'normal',
                                    overflowX: 'hidden',
                                    overflowY: 'hidden',
                                    color: 'transparent',
                                    caretColor: 'white',
                                    resize: 'none',
                                }, value: yaml, onChange: (e) => {
                                    if (!e.target.value) {
                                        setYaml('');
                                        props.setData?.({});
                                    }
                                    else {
                                        setYaml(e.target.value);
                                        try {
                                            const data = YamlToObject(e.target.value, props.isYamlArray);
                                            props.setData?.(data);
                                            setError('');
                                            setErrorLine(-1);
                                        }
                                        catch (err) {
                                            let message = err.message ?? 'Unkown error';
                                            const index = message.indexOf('at line');
                                            if (index !== -1) {
                                                let lineString = message.substring(index).split(' ')[2];
                                                lineString = lineString.slice(0, lineString.length - 1);
                                                const line = Number(lineString);
                                                if (Number.isInteger(line))
                                                    setErrorLine(line - 1);
                                                message = message.substring(0, index);
                                            }
                                            setError(message);
                                        }
                                    }
                                } }, void 0)] }, void 0) }, void 0) }, void 0)] }, void 0));
}
